ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctKudaf.java

```diff
@@ -20,115 +20,104 @@
-    this.comparator = (v1, v2) -> {
-      if (v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
-
-      return Comparator.<T>reverseOrder().compare(v1, v2);
-    };
   }
 
   @Override
-  public T[] aggregate(final T currentValue, final T[] aggregateValue) {
+  public List<T> aggregate(final T currentValue, final List<T> aggregateValue) {
+
     if (currentValue == null) {
       return aggregateValue;
     }
 
-    final T last = aggregateValue[aggregateValue.length - 1];
-    if (last != null && currentValue.compareTo(last) <= 0) {
+    final int currentSize = aggregateValue.size();
+    if (currentSize == tkVal && currentValue.compareTo(aggregateValue.get(currentSize - 1)) <= 0) {
       return aggregateValue;
     }
 
-    if (ArrayUtil.containsValue(currentValue, aggregateValue)) {
+    if (aggregateValue.contains(currentValue)) {
       return aggregateValue;
     }
 
-    final int nullIndex = ArrayUtil.getNullIndex(aggregateValue);
-    if (nullIndex != -1) {
-      aggregateValue[nullIndex] = currentValue;
-      Arrays.sort(aggregateValue, comparator);
-      return aggregateValue;
+    if (currentSize == tkVal) {
+      aggregateValue.set(currentSize - 1, currentValue);
+    } else {
+      aggregateValue.add(currentValue);
     }
 
-    aggregateValue[aggregateValue.length - 1] = currentValue;
-    Arrays.sort(aggregateValue, comparator);
+    aggregateValue.sort(Comparator.reverseOrder());
     return aggregateValue;
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public Merger<String, T[]> getMerger() {
+  public Merger<String, List<T>> getMerger() {
     return (aggKey, aggOne, aggTwo) -> {
-      final T[] merged = (T[]) Array.newInstance(ttClass, tkVal);
+      final List<T> merged = new ArrayList<>(Math.min(tkVal, aggOne.size() + aggTwo.size()));
 
       int idx1 = 0;
       int idx2 = 0;
       for (int i = 0; i != tkVal; ++i) {
-        final T v1 = idx1 < aggOne.length ? aggOne[idx1] : null;
-        final T v2 = idx2 < aggTwo.length ? aggTwo[idx2] : null;
+        final T v1 = getNextItem(aggOne, idx1);
+        final T v2 = getNextItem(aggTwo, idx2);
 
-        final int result = comparator.compare(v1, v2);
-        if (result < 0) {
-          merged[i] = v1;
-          idx1++;
-        } else if (result == 0) {
-          merged[i] = v1;
+        if (v1 == null && v2 == null) {
+          break;
+        }
+
+        if (v1 != null && (v2 == null || v1.compareTo(v2) > 0)) {
+          merged.add(v1);
           idx1++;
+        } else if (v1 == null || v2.compareTo(v1) > 0) {
+          merged.add(v2);
           idx2++;
-        } else {
-          merged[i] = v2;
+        } else if (v1.compareTo(v2) == 0) {
+          merged.add(v1);
+          idx1++;
           idx2++;
         }
       }
-
       return merged;
     };
   }
 
+  private T getNextItem(final List<T> aggList, int idx) {
```

Comment:
lol.  probably a little overkill to have this as a separate function. No biggie, but mark it as static if you want to keep it.

Reply:
Had to move this here to fix the `cyclomatic complexity` error in checkstyle. Cannot make it static since cannot access this for finding out `T`.

Reply:
```
private static <T> T getNextItem(final List<T> aggList, int idx) {
```

Reply:
```
private static <T> T getNextItem(final List<T> aggList, int idx) {
```